src/software/ai/hl/stp/tactic/patrol_tactic_test.cpp

```diff
@@ -0,0 +1,305 @@
+    Point patrol_point2 = Point(6, 9);
+    std::vector<Point> patrol_points{patrol_point1, patrol_point2};
+
+    double at_patrol_point_tolerance = 0.5;
+    double speed_at_patrol_points    = 2.0;
+
+    Robot robot = Robot(robot_id, robot_starting_point, initial_velocity, Angle::zero(),
+                        AngularVelocity::zero(), Timestamp::fromSeconds(0));
+
+    PatrolTactic tactic = PatrolTactic(patrol_points, at_patrol_point_tolerance,
+                                       Angle::zero(), speed_at_patrol_points);
+
+    tactic.updateRobot(robot);
+
+    generateExpectedActions(robot, patrol_points, Angle::zero(), speed_at_patrol_points);
+    generateRobotStatesToCompleteAction(patrol_points, initial_velocity);
+
+    // Act and Assert
+
+    for (int j = 0; j < 2; j++)
+    {
+        for (int i = 0; i < patrol_points.size(); i++)
+        {
+            std::shared_ptr<Action> action_ptr = tactic.getNextAction();
+            ASSERT_NE(nullptr, action_ptr);
+
+            std::shared_ptr<MoveAction> move_action =
+                std::dynamic_pointer_cast<MoveAction>(action_ptr);
+            compareMoveActions(expected_actions[i], move_action);
+
+            simulateActionToCompletion(robot, robot_state_to_complete_actions[i], action_ptr,
+                                       tactic);
+        }
+    }
+}
+
+TEST_F(PatrolTacticTest, cost_of_non_assigned_robot_when_tactic_is_assigned)
+{
+    // setup
+    Point robot_starting_point = Point(1, 1);
+    Vector initial_velocity   = Vector(0, 1);
+
+    Robot assigned_robot = Robot(0, robot_starting_point, initial_velocity, Angle::zero(),
+                                 AngularVelocity::zero(), Timestamp::fromSeconds(0));
+
+    Robot non_assigned_robot =
+        Robot(1, robot_starting_point, initial_velocity, Angle::zero(),
+              AngularVelocity::zero(), Timestamp::fromSeconds(0));
+
+    std::vector<Point> patrol_points{Point(2, 3), Point(4, 3)};
+    World world = ::Test::TestUtil::createBlankTestingWorld();
+
+    PatrolTactic tactic = PatrolTactic(patrol_points, 1.0, Angle::zero(), 2.0);
+
+    tactic.updateRobot(assigned_robot);
+
+    double cost = tactic.calculateRobotCost(non_assigned_robot, world);
+
+    EXPECT_EQ(COST_OF_NONASSIGNED_ROBOT_UNASSIGNED_TACTIC, cost);
+}
+
+TEST_F(PatrolTacticTest, cost_of_non_assigned_robot_when_tactic_is_not_assigned)
+{
+    // setup
+    Point robot_starting_point = Point(1, 1);
+    Vector initial_velocity   = Vector(0, 1);
+
+    Robot non_assigned_robot =
+        Robot(1, robot_starting_point, initial_velocity, Angle::zero(),
+              AngularVelocity::zero(), Timestamp::fromSeconds(0));
+
+    std::vector<Point> patrol_points{Point(2, 2), Point(4, 3)};
+    World world = ::Test::TestUtil::createBlankTestingWorld();
+
+    PatrolTactic tactic = PatrolTactic(patrol_points, 1.0, Angle::zero(), 2.0);
+
+    double cost = tactic.calculateRobotCost(non_assigned_robot, world);
+
+    EXPECT_NEAR(0.15, cost, 0.05);
```

Comment:
0.15 here seems to be a bit magical to me. Just wondering why the cost is compared with 0.15 here?

Reply:
For a patrol tactic that is unassigned the cost of assigning a robot to the tactic is given by 

        `double cost = dist / world.field().totalXLength();`

at the bottom of PatrolTactic::calculateRobotCost().
 
`dist` is the distance between the robots current position (robotStartingPoint (1,1)) and the first patrol point (2,2), which is sqrt(2). 
`world.field().totalXLength();` is something set by the `test::TestUtil::createBlankTestingWorld();`  based on a real fields x length and some buffer. Its value is 9.6. 
Dividing the two values gives 0.1473 ~= 0.15

Hope thats a good enough answer :)

Reply:
make sense, Thanks!